### PR TITLE
[DOC] scientific reference - `Series` and `Panel` types

### PR DESCRIPTION
**Pull request by fkiraly**
This PR addresses issue #1781, and is written in preparation of #1739.

This contains one (currently non-linked) page of scientific reference, which defines "series" and "panels".

Reviewers should concentrate on:
* well-definedness and coherence of mathematical formalism
* agreement with general intuition of time series and panels
* formatting of the rst file

### CODE REVIEW

**Comment by fkiraly** [comment]
FYI @Lovkush-A, would especially appreciate your review (due to your pure math background)

**Comment by fkiraly** [comment]
note to self: we want to add sub-sections with three examples (with different properties)

File: docs/source/scientific_reference/1_time_series.rst [hunk 1]
```diff
@@ -0,0 +1,165 @@
+.. _sciref_series:
+
+Scientific reference: time series
+=================================
+
+``sktime``'s conceptual model is based on a number of "time series" related mathematical
+objects:
+
```
Comment (lmmentel) [inline docs/source/scientific_reference/1_time_series.rst:8]: As a general thought consider using a definition list https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

File: docs/source/scientific_reference/1_time_series.rst [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+Primitive values - the `Primitives` scitype
+-------------------------------------------
+
+The following sets/domains are called basic primitive domains:
+
+* the real numbers, i.e., the set :math:`\mathbb{R}`
+* the set of boolean truth value, i.e., the two-element set containing `True` or `False`
+* the set of all finite length strings of ASCII characters
+* any other finite set :math:`\mathcal{C}`, called "categoriy set" in this context
```
Comment (lmmentel) [inline docs/source/scientific_reference/1_time_series.rst:32]:
```suggestion
* any other finite set :math:`\mathcal{C}`, called "category set" in this context
```

File: docs/source/scientific_reference/1_time_series.rst [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+A set/domain is called primitive domain if it is a basic primitive domain, or
```
Comment (lmmentel) [inline docs/source/scientific_reference/1_time_series.rst:34]: :mage: cannot say that I'm able to follow. Is this user facing documentation or internal dev documentation?

File: docs/source/scientific_reference/1_time_series.rst [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+=================================
+
+``sktime``'s conceptual model is based on a number of "time series" related mathematical
+objects:
+
+* "time series", observations of a single process made over time.
+Example: average temperature in London, on a given day.
+* "time series panels", collections of multiple time series, independently observed.
+Example: fever curves of 10 specific hospital patients, for the time in the hospital.
+* "hierarchical time series", collections of multiple time series where individual
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:13]: panel and hierarchical are commonly used terms. The intuition is usually the same, though the precise formal meaning that authors assign to it differs across sources.

File: docs/source/scientific_reference/1_time_series.rst [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+* scientific types (scitype) - ADT plus statistical assumptions
+
+
+Primitive values - the `Primitives` scitype
+-------------------------------------------
+
+The following sets/domains are called basic primitive domains:
+
+* the real numbers, i.e., the set :math:`\mathbb{R}`
+* the set of boolean truth value, i.e., the two-element set containing `True` or `False`
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:30]: is this really better? If a person only half remembers what a boolean is?

File: docs/source/scientific_reference/1_time_series.rst (hunk 3 above)
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:34]: user facing, for users who are familiar with math style formalism.
This is a formal ADT definition, so will involve mathematical definitions.

Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:34]: a great introduction into this kind of language is "Daepp U, Gorkin P (2011) - Reading, Writing, and Proving (2nd ed)"

Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:34]: the specific lines, btw, define the domain of a data frame row

Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:34]: the doc is meant to complement the API reference

**Comment by fkiraly** [comment]
Thinking about your comments, @lmmentel.
This should be useful for users in general, the current structure might restrict it too much to those with formal math background.

Would it be better if we structure it as follows, for each object introduced:
* intuition
* 3 examples
* formal definition
* notes

instead of starting (and potentially intimidating the reader) with formal definition?

File: docs/source/scientific_reference/1_time_series.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (lmmentel) [reply docs/source/scientific_reference/1_time_series.rst:13]: I'm not sure I understand the distinction of panel and hierarchical based purely on the explanation and example. Maybe you can quote some sources?

File: docs/source/scientific_reference/1_time_series.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (lmmentel) [reply docs/source/scientific_reference/1_time_series.rst:30]: If this is for a person who half remembers what a boolean is it's worth reconsidering all the math formalism.

File: docs/source/scientific_reference/1_time_series.rst (hunk 3 above)
Comment (lmmentel) [reply docs/source/scientific_reference/1_time_series.rst:34]: I'm only saying that it reads unnecessarily formal and opaque, maybe saying "row of a data frame" is ok. I don't immediately see any value in introducing this new terminology.

**Comment by lmmentel** [comment]
>Thinking about your comments, @lmmentel.
This should be useful for users in general, the current structure might restrict it too much to those with formal math background.

>Would it be better if we structure it as follows, for each object introduced:

>intuition
>3 examples
>formal definition
>notes
>instead of starting (and potentially intimidating the reader) with formal definition?

Yeah that seems more accessible, especially if you plan to connect it to the knowledge that users might have about vectors, arrays, series, dataframes matrices etc.

**Review by Lovkush-A — COMMENTED** [review]
definitely useful to have this written down somewhere.

1. reading immentels comments really hits home how unnatural pure mathematical language is for most people! definitely need to have a non-mathematician reviewing this to make sure it is reasonably accessible.
2. need to somehow make clear the difference between a dataframe with index 0,...,n representing a time series and one representing 'tabular data' (where the order of the rows does not matter)
3. right now, as the definitions are written, permuting the columns in a multivariate time series would give you a different time series. (the domains X1 x X2 and X2 x X1 are mathematically distinct if X1 neq X2). I think this is incorrect; we should be able to change the order of the columns.
4. do we need to say anything about how Series can represent other kinds of ordered data, not just time series?

one final note. i think it will be very hard to make this 'perfect' (i.e. both accessible and 100% precise), but i am very much in favour of having something written down somewhere.

**Comment by fkiraly** [comment]
> do we need to say anything about how Series can represent other kinds of ordered data, not just time series?

Well, am I not doing that? I'm introducing "indexed series" in general, where the index can be anything, and "time series" is the sub-case where we have an index that is interpreted as time, which means it must be a sub-set of the (one-dimensional) reals.

**Comment by fkiraly** [comment]
> need to somehow make clear the difference between a dataframe with index 0,...,n representing a time series and one representing 'tabular data' (where the order of the rows does not matter)

The abstract type for tabular data would be different. You would have equivalence up to permutation of rows, which in the case of a time series you do not have. Perhaps adding this as a note?

It might be confusing to most people, since "equivalence up to permutation of rows" is not a conceptual element that they would be aware of, since in the computer, they are looking at actual, distinct rows.

File: docs/source/scientific_reference/1_time_series.rst [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+For a panel of indexed series :math:`p`, with index domain :math:`\mathcal{T}`, we say that
+:math:`p` is:
+
+* a panel of time series, if all series in the panel :math:`p` are time series
+* of homogeneous value domain if all series in the panel :math:`p`
+have the same value domain; that is, if :math:`p_i\in \mbox{panel}(\mathcal{X}, \mathcal{T})`
+for some :math:`\mathcal{X}` and all :math:`i = 1, \dots, N`.
+* uni-/multivariate, if :math:`p` is of homogeneous value domain and all series in the
+panel :math:`p` are uni-/multivariate
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:148]: not of homogeneous value domain, according to the list.

File: docs/source/scientific_reference/1_time_series.rst [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+* defining time series only over a finite index set :math:`t_1,\dots, t_T`,
+instead of, say, defining :math:`s(t)` for all :math:`t` in an infinite or even
+uncountably infinite set (e.g., the reals). This is because, in a computer,
+time series observations are always finite.
+Some books do not make this distinction and conflate "generative" knowledge
+in mathematical assumptions with observed information.
+We made this decisions since real observations, including those that ``python`` can handle, are always finite.
+* considering the index set as part of the time series and crucial, inspectable information.
+Some sources conceptualize the index set as "external" to the time series, or irrelevant.
+We think that considering the index set as part of the time series is conceptually natural for any user of ``pandas``.
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:111]: Yes, since ``pandas`` is so widely used that it is highly influential in formation of concepts.
What I want to say is that it is not the main *reason*, but users of ``pandas`` will be highly familiar with the concept - in contrast to people who read a lot of theory books but are not familiar with ``pandas`` (or R data frames).

File: docs/source/scientific_reference/1_time_series.rst [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+* defining time series only over a finite index set :math:`t_1,\dots, t_T`,
+instead of, say, defining :math:`s(t)` for all :math:`t` in an infinite or even
+uncountably infinite set (e.g., the reals). This is because, in a computer,
+time series observations are always finite.
+Some books do not make this distinction and conflate "generative" knowledge
+in mathematical assumptions with observed information.
+We made this decisions since real observations, including those that ``python`` can handle, are always finite.
+* considering the index set as part of the time series and crucial, inspectable information.
+Some sources conceptualize the index set as "external" to the time series, or irrelevant.
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:110]: I think it's simply outside their conceptual domain.

File: docs/source/scientific_reference/1_time_series.rst [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+We use the scitype `Primitives` to type objects of primitive type.
+
+
+Time Series - the `Series` scitype
+----------------------------------
+
+An indexed series, with value domain :math:`\mathcal{X}`, index domain :math:`\mathcal{T}`,
+of length :math:`T`, is a collection of index points :math:`t_1, \dots, t_T` and observations
+:math:`x_1, \dots x_T`. We say (and interpret) that :math:`x_i` is the value observed at index
+:math:`t_i`.
```
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:48]: the challenge is that "being a time series" is not a property of the object, it is a matter of typing (attaching interpretation). Since a real index does not necessarily have to be "time". Could be "distance" or "wavelength", as in common "time" series classification applications.

**Comment by Lovkush-A** [comment]
> I disagree! The ordering of the variables matters, in my opinion and personal conceptual model.

* There are classes/functions in which we automatically rearrange the columns to match what was seen during fit. To me, this means that we consider permutations of variables/columns as symmetries of the scitype.
* I do not see why one would want the order of the variables as a part of the conceptual model. (except maybe to make definitions easier to write down...). For me, it is additional structure that one is keeping track of, which has no scientific meaning. Do you know any scientific questions that change if you swap two columns around? Forecasting multivariate series with cols A,B,C is the same question to forecasting B,A,C, as far as I know.

> I was wondering whether we want named variables - i.e., should we include _column index_ in the definition, or not?

I think so. Especially since pandas dataframes are the 'foundational' datatype in our codebase (as opposed to numpy arrays, say)

**Comment by Lovkush-A** [comment]
> > do we need to say anything about how Series can represent other kinds of ordered data, not just time series?
> 
> Well, am I not doing that?.

No. To quote: "We use the scitype Series to type univariate and multivariate time series"

> I'm introducing "indexed series" in general, where the index can be anything, and "time series" is the sub-case where we have an index that is interpreted as time, which means it must be a sub-set of the (one-dimensional) reals.

I dont know if I am being pedantic, but we do need to say something about how with *time* series there is assymettry between past and future, but in other kinds of real-indexed series (eg the arrow head data) there might not be the assymetry.

**Comment by Lovkush-A** [comment]
> The abstract type for tabular data would be different. You would have equivalence up to permutation of rows, which in the case of a time series you do not have. Perhaps adding this as a note?

I think we should add this as a note. It is precisely this property (i think) that distinguishes between tabular data and time series data. (random thought...is it possible to have sklearn but where fundamental objects are sets of rows, rather than ordered lists of rows. 🤔 )

> It might be confusing to most people, since "equivalence up to permutation of rows" is not a conceptual element that they would be aware of

They might not be aware of it in those terms (in the same way most users would be confused by the rest of the document). However, I presume they are used to the fact that in sklearn, if you jumble up the rows, it should not affect one's model.

File: docs/source/scientific_reference/1_time_series.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:13]: a panel is just a collection, whereas a hierarchical time series has hierarchy labels attached to different series.
What complicates the conceptualization is that panels can be seen as a sub-case of hierarchical, where the hierarchy label is a (flat) set of categories or integer indices.###

File: docs/source/scientific_reference/1_time_series.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:30]: (ouch, but point taken)

**Comment by fkiraly** [comment]
> > > do we need to say anything about how Series can represent other kinds of ordered data, not just time series?
>> 
>> Well, am I not doing that?.
>
>No. To quote: "We use the scitype Series to type univariate and multivariate time series"

Ok, this is going to get subtle.

In `sktime`, so far, we assume that all indices are *time*, or at least time-like, i.e., indexed series with an index supported in the reals. The `Series` scitype is a type attached to univariate and multivariate time series. Should I instead say, univariate and multivariate indexed series with a time-like index domain? That would be quite confusing, although more precise.

>> I'm introducing "indexed series" in general, where the index can be anything, and "time series" is the sub-case where we have an index that is interpreted as time, which means it must be a sub-set of the (one-dimensional) reals.
>
> I dont know if I am being pedantic, but we do need to say something about how with *time* series there is assymettry between past and future, but in other kinds of real-indexed series (eg the arrow head data) there might not be the assymetry.

That there is assymmetry is mirrored by the *absence* of the assumption that we have symmetry - is that not good enough? Perhaps the examples should point this out?

**Comment by fkiraly** [comment]
> (random thought...is it possible to have sklearn but where fundamental objects are sets of rows, rather than ordered lists of rows. 🤔 )

The best concrete data type for that would still be the `pd.DataFrame`, no? With the interface convention that all operations carried out must be invariant under row permutations of individual data containers?

**Comment by Lovkush-A** [comment]
> Ok, this is going to get subtle.

Yes it is!  I am confused by what the term "time series" means in your formalisation. Quoting the document:
    - "may choose to say that s is a time series if \mathcal{T} is totally ordered and (order)-isomorphic to a sub-set of \mathbb{R}, in a **context where elements of \mathcal{T} are interpreted as time stamps**"

* is the arrowhead dataset a time series, under your conceptualisation? if yes, then I think the definition above is inaccurate, because the index in arrowhead is not interpreted as time. (or is it?? i dont think it should be...)

> Should I instead say, univariate and multivariate indexed series with a time-like index domain? That would be quite confusing, although more precise.

* Why is this confusing? To me, saying the arrow head dataset is a time series is confusing (given definitions above). 'Time-like' seems understandable.
* if we are going to write formal mathematical definitions, presumably the aim is precision. otherwise, I dont see who this benefits. mathematicians (like me) will be confused and most non-mathematicians will skip the formalities and just use the examples and intuition anyway.

> That there is (potential) asymmetry in an object satisfying the definition is mirrored by the _absence_ of the assumption that we have symmetry - is that not good enough? Perhaps the examples should point this out?

I think this depends on what you want 'time series' to mean. For me, *if* time series index really correspond to time stamps, then the assymetry is implicitly a part of the definition and so we should explicitly mention it.

File: docs/source/scientific_reference/1_time_series.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply docs/source/scientific_reference/1_time_series.rst:48]: i am not sure how your comment is a response to my comment...?

[the remark this answers is not in the document]
Comment (fkiraly) [reply docs/source/scientific_reference/1_time_series.rst:48]: It's an answer in that option 1 is not well-defined given the framework.

> and then define a time series as a particular kind of indexed series"

Not an option because: "being a time series" is not a property of an indexed series. At least not how it is defined currently.
Therefore, a time series cannot be a "particular kind of" indexed series, because that would have to be defined by a property.

Why is it not a property? Say, you define a time series to be of real domain. As stated above, the reals could also be "wavelength" or "distance".

File: docs/source/scientific_reference/1_time_series.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (TonyBagnall) [reply docs/source/scientific_reference/1_time_series.rst:13]: I don't really understand the difference between a multivariate instance and a hierarchical instance, and why that distinction is important. 

1. Is the latter is meant to include relationship data (e.g. spatial data)? If so, is there a mechanism for encoding this? A EEG is under this interpretation hierarchical, as is uman activity recognition, but afaik we do not have a standard for encoding this within the data structure, so the difference is not in the implementation. 
2. Or do we allow different sampling processes/start times for hierarchical? What actually is the "hierarchy"? If so, then it would have to be stored as a pd.Series, and that could be a valid distinction. 
3. Or are some series actually derived from others? Such as total sals = sum of all sales at the next level of hierarchy?

File: docs/source/scientific_reference/1_time_series.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply docs/source/scientific_reference/1_time_series.rst:48]: I do not fully understand the nuance here.
1. How about this wording instead: 'we first define indexed series and then time series'. Or, have a contents section at the top of the page which says that you define indexed series and time series in the section. Or some other way to indicate that we are defining two different things in this section.
2. Is the following statement true: 'all time series are indexed series'. If no - then I am confused. If yes, then this is the sense in which I mean that time series are a 'particular kind' of indexed series.
